### DISCUSSION TITLE
Life-cycle of providers

### DISCUSSION THREAD

**Discussion by najibghadri** [rrousselGit/riverpod#1029] [Q&A] [answered]
Hello there!

I noticed that even though we define providers and their dependencies globally, they are not created/initialized/constructed until they are either `read`, `listen`ed or `watch`ed once from a `build`. Although then they will be created in the right dependency order indeed, if there are providers that depend on others. However my problem with this is that provider creation is thus tied to UI.

To me providers represent the backend logic for UI. For example in some of my provider constructors, in my chat app, I kick-off async `load` functions that will load the contacts and the conversations for the current user. I want to do these before showing any of the contact or conversation pages... To do this I found myself calling `ref.read(provider)` in the `build` of a top-most component, to initialize that provider.

What do you think about this? Am I using (riverpod) providers completely wrong or is this a valid concern?
Thank you and Happy Holidays! 🎆

**Comment by rrousselGit** [comment, marked as answer]
Reading the providers that you want to eagerly initialize in the top-most components is the correct solution

Due to how Dart works, there's currently no way to do otherwise. Maybe that'll change with metaprogramming, but for now that'll be your only solution

**Reply by najibghadri** [reply]
Thank you @rrousselGit ! 🙏
